File: 2542-maximum-subsequence-score/2542-maximum-subsequence-score.swift
class Solution {
    func maxScore(_ nums1: [Int], _ nums2: [Int], _ k: Int) -> Int {
        let z = zip(nums1, nums2).sorted { $0.1 < $1.1 }
        let n = nums1.count
        var minHeap = [Int]()
        var maxScore = -1
        var sum = 0
        for i in (0 ..< n).reversed() {
            heappush(&minHeap, z[i].0)
            sum += z[i].0
            if minHeap.count > k {
                sum -= heappop(&minHeap)
            }
            if (n-i) >= k {
                maxScore = max(maxScore, sum * z[i].1)
            }
        }
        return maxScore
    }
    
    private func heappush(_ heap: inout [Int], _ element: Int) {
        heap.append(element)
        var currentIndex = heap.count - 1
        var parentIndex = (currentIndex - 1) / 2
        while parentIndex >= 0 && heap[parentIndex] > heap[currentIndex] {
            heap.swapAt(parentIndex, currentIndex)
            currentIndex = parentIndex
            parentIndex = (currentIndex - 1) / 2
        }
    }
    
    private func heappop(_ heap: inout [Int]) -> Int {
        heap.swapAt(0, heap.count - 1)
        let ret = heap.removeLast()
        var currentIndex = 0
        var leftChildIndex = 1
        var rightChildIndex = 2
        while leftChildIndex < heap.count {
            var minIndex = currentIndex
            if heap[leftChildIndex] < heap[minIndex] {
                minIndex = leftChildIndex
            }
            if rightChildIndex < heap.count && heap[rightChildIndex] < heap[minIndex] {
                minIndex = rightChildIndex
            }
            if minIndex == currentIndex {
                break
            }
            heap.swapAt(currentIndex, minIndex)
            currentIndex = minIndex
            leftChildIndex = currentIndex * 2 + 1
            rightChildIndex = currentIndex * 2 + 2
        }
        return ret
    }
}

File: 2542-maximum-subsequence-score/2542-maximum-subsequence-score.py
class Solution:
    def maxScore(self, nums1: List[int], nums2: List[int], k: int) -> int:
        z = sorted(zip(nums1, nums2), key=lambda x: x[1])
        n = len(nums1)
        minHeap = []
        maxScore = -1
        total_sum = 0

        for i, el in enumerate(reversed(z)):
            heapq.heappush(minHeap, el[0])
            total_sum += el[0]
            if len(minHeap) >= k:
                maxScore = max(maxScore, total_sum * el[1])
                total_sum -= heapq.heappop(minHeap)

        return maxScore


s = Solution()
f = open("user.out", "w")

for nums1, nums2, k in zip(map(loads, stdin), map(loads, stdin), map(loads, stdin)):
    f.write(dumps(s.maxScore(nums1, nums2, k)).replace(" ", "") + "\n")
exit()
